ocaml/database/database_test.ml

```diff
@@ -0,0 +1,691 @@
+    if (List.mem_assoc "VBDs" vm_updated)
+    then (Printf.printf "Pass\n")
+    else (Printf.printf "Fail\n"; failwith "Event problem");
+
+    let (_: unit) = Client.write_field t "VBD" vbd "VM" "moo" in
+    let db = Db_ref.get_database t in
+    let g5 = get_max db in
+    Printf.printf "generation after write_field is: %Ld\n" g5;
+    dump db g4;
+    let updated = get_updated db g4 in
+    Printf.printf "===TEST=== Checking one-to-many after one-update: ";
+    let vm_updated = List.filter (fun (r,_) -> r=vm) updated in
+    let vm_updated = List.map snd vm_updated in
+    if (List.mem_assoc "VBDs" vm_updated)
+    then (Printf.printf "Pass\n")
+    else (Printf.printf "Fail\n"; failwith "Event problem");
+
+    let (_: unit) = Client.write_field t "VBD" vbd "type" "Banana" in
+    let db = Db_ref.get_database t in
+    let g6 = get_max db in
+    Printf.printf "generation after write_field is: %Ld\n" g6;
+    dump db g5;
+    let updated = get_updated db g5 in
+    Printf.printf "===TEST=== Checking one-to-many after one-update of non-reference field: ";
+    let vm_updated = List.filter (fun (r,_) -> r=vm) updated in
+    let vm_updated = List.map snd vm_updated in
+    if not (List.mem_assoc "VBDs" vm_updated)
+    then (Printf.printf "Pass\n")
+    else (Printf.printf "Fail\n"; failwith "Event problem");
+
+    let (_ : unit) = Client.delete_row t "VBD" vbd in
+    let db = Db_ref.get_database t in
+    let g7 = get_max db in
+    Printf.printf "generation after delete VBD is: %Ld\n" g7;
+    Printf.printf "===TEST=== Checking deleted event: ";
+    let deleted = get_deleted db g6 in
+    if (List.mem vbd deleted)
+    then (Printf.printf "Pass\n")
+    else (Printf.printf "Fail\n"; failwith "Event problem");
+
+    Client.create_row t "VBD" (make_vbd vm vbd vbd_uuid) vbd;
+    let (_ : unit) = Client.delete_row t "VBD" vbd in
+    let db = Db_ref.get_database t in
+    let g8 = get_max db in
+    Printf.printf "generation after create/delete VBD is: %Ld\n" g8;
+    Printf.printf "===TEST=== Checking the VBD doesn't appear in the deleted list: ";
+    let deleted = get_deleted db g7 in
+    if not (List.mem vbd deleted)
+    then (Printf.printf "Pass\n")
+    else (Printf.printf "Fail\n"; failwith "Event problem");
+    dump db g7;
+
+    Client.create_row t "VBD" (make_vbd vm vbd vbd_uuid) vbd;
+    let db = Db_ref.get_database t in
+    let g9 = get_max db in
+    let (_ : unit) = Client.delete_row t "VBD" vbd in
+    Client.create_row t "VBD" (make_vbd vm vbd2 vbd_uuid2) vbd2;
+    let (_ : unit) = Client.delete_row t "VBD" vbd2 in
+    let db = Db_ref.get_database t in
+    let g10 = get_max db in
+
+    Printf.printf "===TEST=== Checking for masking of delete events: ";
+
+
+    let deleted = get_deleted db g9 in
+    if (List.mem vbd deleted)
+    then (Printf.printf "Pass\n")
+    else (Printf.printf "Fail\n"; failwith "Event problem");
+    dump db g9;
+    ignore(g10);
+
+
+
+    ()
```

Comment:
The choice of spacing and newlines are "interesting" in some part of this code :stuck_out_tongue_winking_eye:

Reply:
Ah, I've forgotten to ocp-indent it.